### PR TITLE
refactor(behavior_path_planner): refactor planWithPriority in start_planner

### PR DESCRIPTION
**Pull request by kyoichi-sugahara**
## Description
minor refactor the function `planWithPriotity`
the [PR](https://github.com/autowarefoundation/autoware.universe/pull/5390) should be merged.

---

### <samp>🤖 Generated by Copilot at 6ccd8a2</samp>

This pull request enhances the start planner module, which plans a path for the ego vehicle to pull out of a parking spot. It improves the readability and maintainability of the code by introducing type aliases, renaming fields and methods, and refactoring the logic with new private methods. It also changes the compatibility check with other scene modules based on the vehicle's direction.

## Tests performed
- [x] [PASS TIERIV INTERNAL SCENARIOS](https://evaluation.tier4.jp/evaluation/reports/f3be1a43-5c52-5634-9e30-14d0ed7cbc3f?project_id=prd_jt) 

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp [hunk 1]
```diff
@@ -492,102 +492,112 @@ void StartPlannerModule::planWithPriority(
...
-      status_.pull_out_path = *pull_out_path;
-      status_.pull_out_start_pose = pull_out_start_pose;
-      status_.planner_type = planner->getPlannerType();
-      return true;
-    }
+  if (start_pose_candidates.empty()) return;
 
-    // If this is the last start pose candidate, return false
-    if (i == start_pose_candidates.size() - 1) return false;
+  PriorityOrder order_priority =
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp:497]:
```suggestion
  const PriorityOrder order_priority =
```

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp [hunk 2]
```diff
@@ -492,99 +492,109 @@ void StartPlannerModule::planWithPriority(
...
-    const auto pull_out_path_next = planner->plan(pull_out_start_pose_next, goal_pose);
-    // not found safe path
-    if (!pull_out_path_next) {
-      return false;
-    }
+  for (const auto & pair : order_priority) {
+    if (findPullOutPath(
+          start_pose_candidates, pair.first, pair.second, refined_start_pose, goal_pose))
+      return;
+  }
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp:504]: I prefer
```suggestion
  for (const auto & [index, planner] : order_priority) {
    if (findPullOutPath(
          start_pose_candidates, index, planner, refined_start_pose, goal_pose))
      return;
  }
```

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp [hunk 3]
```diff
@@ -492,99 +492,109 @@ void StartPlannerModule::planWithPriority(
...
-    return true;
-  };
+  updateStatusIfNoSafePathFound();
+}
 
-  using PriorityOrder = std::vector<std::pair<size_t, std::shared_ptr<PullOutPlannerBase>>>;
-  const auto make_loop_order_planner_first = [&]() {
-    PriorityOrder order_priority;
+PriorityOrder StartPlannerModule::determinePriorityOrder(
+  const std::string & search_priority, size_t candidates_size)
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp:510]:
```suggestion
  const std::string & search_priority, const size_t candidates_size)
```

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp [hunk 4]
```diff
@@ -492,99 +492,109 @@ void StartPlannerModule::planWithPriority(
...
+  const behavior_path_planner::PullOutPath & path, const Pose & start_pose,
+  const std::shared_ptr<PullOutPlannerBase> & planner)
+{
+  const std::lock_guard<std::mutex> lock(mutex_);
+  status_.driving_forward = false;
+  status_.found_pull_out_path = true;
+  status_.pull_out_path = path;
+  status_.pull_out_start_pose = start_pose;
+  status_.planner_type = planner->getPlannerType();
+}
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp:594]: nits:
giving planner_type instead of planner is enough

Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp:594]: I agree! thanks!

Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp]: changed in https://github.com/autowarefoundation/autoware.universe/pull/5393/commits/791145e6f8368693f6ed74cfa19b6fb33a7a7434

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp (hunk 3 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp]: changed in https://github.com/autowarefoundation/autoware.universe/pull/5393/commits/791145e6f8368693f6ed74cfa19b6fb33a7a7434

File: planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp (hunk 2 above)
Comment (kyoichi-sugahara) [reply planning/behavior_path_planner/src/scene_module/start_planner/start_planner_module.cpp]: I didn't know this way and looks better!
changed in https://github.com/autowarefoundation/autoware.universe/pull/5393/commits/791145e6f8368693f6ed74cfa19b6fb33a7a7434

**Review by kosuke55 — APPROVED** [review]
thanks!!

**Pull request merged**